File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
import sys
from collections import deque
input = sys.stdin.readline

def bfs(y, x) :
    visited = [[-1] * l for _ in range(l)]
    q = deque([(y, x)])
    visited[y][x] = 0

    while q :
        y, x = q.popleft()

        if (y, x) == (target_y, target_x) :
            return visited[y][x]

        for dy, dx in directions :
            ny = y + dy
            nx = x + dx

            if 0 <= ny < l and 0 <= nx < l :
                if visited[ny][nx] == -1 :
                    visited[ny][nx] = visited[y][x] + 1
                    q.append((ny,nx))

    return "갈수없음"

T = int(input())

directions = [(-1, -2),(-2,-1),(-2, 1),(-1, 2),(1,2),(2,1),(1,-2),(2,-1)]

for _ in range(T):
    # 입력 및 초기화
    l = int(input())
    cy, cx = map(int,input().split())
    target_y, target_x = map(int, input().split())

    # bfs
    print(bfs(cy,cx))

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

class Node {
    int x, y, dist;

    Node(int x, int y, int dist) {
        this.x = x;
        this.y = y;
        this.dist = dist;
    }
}

public class Main {
    static int I;
    static int[][] board;
    static boolean[][] visited;

    static int[] dx = {1, 2,-1, -2, 2, 1, -2, -1};
    static int[] dy = {2, 1, 2, 1, -1, -2, -1, -2};
    static int result;

    static int targetX, targetY;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            I = Integer.parseInt(br.readLine());

            board = new int[I][I];
            visited = new boolean[I][I];


            StringTokenizer st = new StringTokenizer(br.readLine(), " ");

            int startX = Integer.parseInt(st.nextToken());
            int startY = Integer.parseInt(st.nextToken());

            st = new StringTokenizer(br.readLine(), " ");
            targetX = Integer.parseInt(st.nextToken());
            targetY = Integer.parseInt(st.nextToken());

            System.out.println(bfs(new Node(startX, startY,0)));
        }
    }

    private static int bfs(Node node) {
        Queue<Node> q = new LinkedList<>();
        q.add(node);
        visited[node.x][node.y] = true;

        while (!q.isEmpty()) {
            Node poll = q.poll();

            int cx = poll.x;
            int cy = poll.y;

            if (cx == targetX && cy == targetY) return poll.dist;

            for (int i = 0; i < 8; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if(nx < 0 || nx >=I || ny < 0 || ny >= I) continue;
                if(visited[nx][ny]) continue;

                visited[nx][ny] = true;
                q.add(new Node(nx, ny, poll.dist + 1));

            }
        }

        return -1;
    }
}